[Autotuner] Plotting utility + test

luarss opened the pull request:
- Plotting utility
- README

luarss committed:
update plotting util and readme

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
```diff
@@ -142,6 +142,18 @@ python3 distributed.py --design gcd --platform sky130hd \
+#### Plot images
+
+After running the autotuner experiments, you can visualize the results as follows.
+Currently, we support the following metrics:
```
```suggestion
After running an AutoTuner experiment, you can generate a graph to understand the results better.
The graph will show one of the metrics listed below progression over the execution of the experiment.
```

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,102 @@
+        "results_dir",
+        type=str,
+        help="Directory containing the results.",
+        default="../../../../../flow/logs/asap7/gcd/test-tune-2024-09-17-12-00-44",
```
```suggestion
```
Remove hardcoded.
This argument should be required.

luarss committed:
make arg required and update docs

Signed-off-by: luarss <jluar@precisioninno.com>

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
```diff
@@ -146,6 +146,19 @@ python3 distributed.py --design gcd --platform sky130hd \
+#### Plot images
+
+After running an AutoTuner experiment, you can generate a graph to understand the results better.
+The graph will show one of the metrics listed below progression over the execution of the experiment.
```
```suggestion
The graph will show the progression of one metric (see list below) over the execution of the experiment.
```

vvbandeira committed:
Update docs/user/InstructionsForAutoTuner.md

Signed-off-by: Vitor Bandeira <vvbandeira@users.noreply.github.com>

vvbandeira reviewed (changes requested):
Please add a call to this script in the `test_helper.py` so we test this script.

luarss committed:
add one call to test plot.py

Signed-off-by: luarss <jluar@precisioninno.com>

luarss committed:
add test_autotuner call, matplotlib reqs

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
temporarily disable unstable tests

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on flow/test/test_autotuner.sh:
```diff
@@ -20,28 +22,32 @@ python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM}SweepSmokeT
+#     python3 distributed.py --design gcd --platform asap7 --server 127.0.0.1 --port 10001 \
+#     --config ../../../../flow/designs/asap7/gcd/autotuner.json tune --samples 1"
+#         docker compose down -v --remove-orphans
+# fi
```
Remove?

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,106 @@
+            params.append(_dict)
+        except Exception as e:
+            print(f"Error in {fname}: {e}")
+            continue
```
Should we keep a list of the files that were not processed so we can print them at the very end in order for the error not to be buried in the logs? And should the script fail silently in this case?

luarss replied:
> And should the script fail silently in this case?

Just to clarify, this exception happens when any Autotuner run does not proceed to the "finish" stage, it does not fail silently

> Should we keep a list of the files that were not processed so we can print them at the very end in order for the error not to be buried in the logs? 

Sure I can add that.

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,106 @@
+    ax.set_ylabel(key)
+    ax.set_title(f"{key} vs Time")
+    z = np.polyfit(df["timestamp"], df[key], 1)
+    p = np.poly1d(z)
```
`z` I assume is the z-axis; what is `p`?

luarss replied:
`z` is the np array of polynomial (ax+b) coefficients, and `p` is the actual polynomial function.

vvbandeira replied:
Then let's use more descriptive names, maybe:
`z` --> `coeff`
`p` --> `poly_func`

luarss committed:
print all files at the end of load_dir

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,109 @@
+    plt.boxplot(df[key])
+    plt.ylabel(key)
+    plt.title(f"{key} Boxplot")
+    plt.savefig(f"images/{key}-boxplot.png")
```
Let's also parametrize the output folder to save images from CI into the reports folder for the design. This way it will be automatically saved to the build artifacts.

luarss committed:
refactors:
- run plot.py for all log files in the autotuner supported design/platforms
- parameterised img_dir, so we can store all AT image runs in build artifacts
- argparse interface now takes in platform, design, experiment for easier usage
- chdir before running plot.py
- add docstrings
- more error handling

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on flow/test/test_autotuner.sh:
```diff
@@ -20,28 +22,23 @@ python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM}SweepSmokeT
+all_experiments=$(basename -a $all_experiments)
+for expt in $all_experiments; do
+  python3 tools/AutoTuner/src/autotuner/utils/plot.py \
+    --platform ${PLATFORM} \
+    --design ${DESIGN_NAME} \
```
At the beginning of this script, you are casting these values to be uppercase, which does not match the path since Linux is case-sensitive. From the CI:
```
02:32:36  Running Autotuner plotting smoke test
02:32:36  ls: cannot access './flow/logs/SKY130HD/gcd/*/': No such file or directory
02:32:36  No experiments found for plotting
02:32:36  + exit 0
```

vvbandeira replied:
If we call the plot script and do not find data to plot, this should be an error, right?

luarss replied:
Yup, these are 2 separate errors, will fix them.

luarss committed:
fix lowercase platform

Signed-off-by: luarss <jluar@precisioninno.com>

luarss committed:
fix syntax for string join

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix LOWERCASE_PLATFORM

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
Change qor key to metric

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on flow/test/test_autotuner.sh:
```diff
@@ -20,28 +22,23 @@ python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM}SweepSmokeT
 python3 -m unittest tools.AutoTuner.test.smoke_test_sample_iteration.${PLATFORM}SampleIterationSmokeTest.test_sample_iteration
 
-if [ "$PLATFORM" == "asap7" ] && [ "$DESIGN" == "gcd" ]; then
+if [ "$PLATFORM" == "asap7" ] && [ "$DESIGN_NAME" == "gcd" ]; then
```
The first check will never be true because you are casting to uppercase on line 14.
Would it be simpler/easier if we just use the platform name everywhere and just remove the dashes if that is not supported by the `unittest` Python lib?

luarss replied:
The main concern is that I prefer name of smoke tests to be capitalized. We could also do something like `asap7SampleIterationSmokeTest`, but might be less readable

Perhaps a `UPPERCASE_PLATFORM` would make the distinction clearer?

vvbandeira replied:
Modifying PLATFORM or another variable already in use somewhere else in the code can easily lead to confusion. Other options could be `PLATFORM_UT` where UT is 'under test' or `TEST_PLATFORM`. That said, I don't see any issue with `asap7SampleIterationSmokeTest`.

vvbandeira commented on flow/test/test_autotuner.sh:
```diff
@@ -20,28 +22,23 @@ python3 -m unittest tools.AutoTuner.test.smoke_test_sweep.${PLATFORM}SweepSmokeT
+all_experiments=$(ls -d ./flow/logs/${LOWERCASE_PLATFORM}/${DESIGN_NAME}/*/)
+if [ -z "$all_experiments" ]; then
+  echo "No experiments found for plotting"
+  exit 0
```
This is still exiting with a success code.

luarss replied:
This is returning 0 because I didn't want to error out for smoke tests that are not suitable for plotting. I will modify it to only run `plot.py` for valid smoke tests then.

vvbandeira replied:
We could then change line 31 to list either manually the tests we want or use a better regex in the experiment names to differentiate whether the plot is available or not.

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,195 @@
+    progress_csvs = glob.glob(f"{dir}/*/progress.csv")
+    if len(progress_csvs) == 0:
+        print("No progress.csv files found.")
+        sys.exit(0)
```
This should also be an error and exit with non-zero value.

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,195 @@
+        print(
+            "Unable to merge DFs due to missing trial_id in params.json (possibly due to failed trials.)"
+        )
+        sys.exit(0)
```
Same as above.

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -0,0 +1,195 @@
+        print(
+            f"KeyError: {e} in the DataFrame. Dataframe does not contain necessary columns."
+        )
+        sys.exit(0)
```
Same as above.

luarss committed:
PLATFORM now strictly lowercase in scripts

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
change plot.py sys exits

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
Fix genReport

Signed-off-by: Jack Luar <jluar@precisioninno.com>
